### ISSUE TITLE
CORS issue with keybase.io API prevents preflight OPTIONS requests via XHR in JavaScript

### ISSUE THREAD

**Issue by JasonCoombs** [keybase/keybase-issues#1959]
Cross-Origin Resource Sharing (CORS) Background, W3C recommendation and historical issues:

https://www.w3.org/TR/cors/#access-control-allow-origin-response-header
https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Preflighted_requests
http://www.html5rocks.com/en/tutorials/cors/
https://bugs.jquery.com/ticket/8423

note: Fetch is the Living Standard meant to replace CORS
https://fetch.spec.whatwg.org/
https://github.com/whatwg/fetch

Summary of Problem:

CORS is awesome. The keybase.io API doesn't allow it, presently. The Access-Control-Allow-Origin HTTP header indicates whether a resource can be shared across domains. For example if one wanted to host keybase.io functionality embedded on one's own website, like this:

https://jasoncoombs.com/verify.php

Then the Allow-Control-Allow-Origin: * HTTP header must be provided by keybase.io in order to permit CORS/Fetch requests which are automated through JavaScript originating from a cross-origin domain such as through the XMLHTTPRequest interface for AJAX scripting. For testing purposes I installed this Chrome extension, which forces the insertion of the HTTP header:

https://chrome.google.com/webstore/detail/allow-control-allow-origi/nlfbmbojpeacfghkpbjhddihlkkiljbi

This partly-resolved the problem, because Chrome stopped refusing to invoke XHR to access the keybase.io API after it started receiving the required HTTP header via the Chrome extension. This made it possible to use the keybase.io fetch API as a CORS request from within sitewide-js:

https://keybase.io/docs/api/1.0/call/key/fetch

However, the keybase.io server refuses to verify preflight OPTIONS requests. It was necessary to modify sitewide-js.js to force XHR to stop preflight. The changes were minor, and anyone can see how this works by using the Chrome extension, above, with the URL of my custom verify.php hosted at jasoncoombs.com but the changes made to hack this together might not be desired in general for other uses of XHR in the keybase web API.

Here are my revised script and the original, plus the diff results after both have been made pretty:

https://keybase.io/_/brew/_/9aa70b7b/sitewide-js.js
https://jasoncoombs.com/sitewide-js.js

$ diff sitewide-js-pretty.js 'sitewide-js original-pretty.js'
13c13
< var baseUrl = "https://keybase.io/";
---
> //# sourceMappingURL=
5500c5500
<             contentType: "text/plain",
---
>             contentType: "application/x-www-form-urlencoded; charset=UTF-8",
5636c5636
<                 jqXHR.setRequestHeader("Content-Type", "text/plain")
---
>                 jqXHR.setRequestHeader("Content-Type", s.contentType)
5939a5940,5942
>                         }
>                         if (!options.crossDomain && !headers["X-Requested-With"]) {
>                             headers["X-Requested-With"] = "XMLHttpRequest"

As you can see, to work around the automated preflight OPTIONS request, which the keybase.io API rejects for some reason, required removing X-Requested-With and forcing Content-Type to "text/plain" in two places.

For testing purposes, my wife sent me a signed love note. Her key is:

https://keybase.io/_/api/1.0/key/fetch.json?pgp_key_ids=4e3d8619f70b333a&ops=4

With my revised script it is possible to paste the following into the Verify page on my server or at keybase.io/verify and your browser's CORS request succeeds when fetching my wife's key from the keybase.io API (remember that the Chrome extension to force CORS to work must be active, until the keybase.io server itself sends the Allow-Control-Allow-Origin header):

-----BEGIN PGP MESSAGE-----
Comment: https://keybase.io/download
Version: Keybase Go 1.0.7 (windows)

xA0DAAoBTj2GGfcLMzoBy+F0AOIAAAAA5EkgbG92ZSB5b3UsIGhvbmXheSEAwsFc
BAABCgAQBQJWkFwRCRBOPYYZ9wszOgAA0tcQAItV6r329owQTvUngYS561fTay6R
kUzbzvq2GKWcTFZ/sp0LoQH0i8HerfDYzWZ3rUCXdWvwggVqupGJEUPFTz2W4910
Iov6xTEs2/isvOHIenfmjx8HT25AdsXSnfpINFkRIJR7imHDfqNbcRQTOR8wRfLU
joh3g+j1VGhSBpFQMVyWhyv2fA6cKm+rAZ0eJu+GcpOVF1R3wJWBh3eisQG9ZU2X
U5xmjzRbPJxbzo66u7KNOJFRlpxGwIt2dqg4d1RIVq2CFIcGZoEJ0w8xtDml7WmX
CVEz5Wy6ffP0n+Rh4MYJ/qN7mERq4mw4j2Qdhotf4/2zS7zT7bFw76IxKcI92Rzd
A5RK/TrZNgC5ClEIPgHDw8rNlrh6cB+AYFyjCMtaqIKgy1aFoL3a5+F9wmmlD5tn
JHPBiIJ7kPpx/GJ2W5gR8WV+iLfkWWUxQ9MGh/QUQueSd6QoXEbvRDtNEN4nY+Ln
dnTC9Zq4AUL4f27Q0TNvM8nTUxPsDSYKG5CV5bAeTVjc7kBKvy8eADn6BsNUjAEk
/SZGGC4CxMRGcu+xaVA1D3BwqvvUxI/zVLRuaDJ1akfsWK0lKSgUamvdwFsoOqYG
V/DtqvxChx+4eQq4sow+jbqK93yRYuHM2i983WgdYzSxRdngwIIWKtp2otfeHSMT
fc6ppa7rAmFsjnST
=Y+Sh
-----END PGP MESSAGE-----

Finally, because this was what I attempted to do in the first place which I found could not be done yet using keybase.io/verify one more line of JavaScript was inserted in my version of sitewide-js to pre-fill the message input from a URL-encoded search string if one is supplied.

The line I've inserted to do this is:

$("#input-msg").val(decodeURI(window.location.search).substr(1));

This makes it possible to turn the PGP MESSAGE block, above, into a clickable hyperlink. There is already a _run_if_prefilled() function defined in sitewide-js.js at keybase.io/verify so presumably it is a feature already on the drawing board. I hope this report helps to show that it is a feature that already works, with the normal caveat that cross-site scripting vulnerabilities need to be properly mitigated in how the keybase.io API receives and responds to user input. In this case the user input is being inserted into the HTML input tag which should make it easier to prevent XSS attacks.

I like this idea because without this functionality the end-user must be clever enough to copy and paste the message in order to verify and read it. Many users are only clever enough to click on links, and on mobile devices NOBODY is clever enough to copy-and-paste.

Try this URL (if you have the Chrome extension installed) and you will instantly see my wife's love note, with verification that she actually wrote it and signed it with her PGP key:

https://jasoncoombs.com/verify.php?-----BEGIN%20PGP%20MESSAGE-----%0D%0AComment:%20https://keybase.io/download%0D%0AVersion:%20Keybase%20Go%201.0.7%20(windows)%0D%0A%0D%0AxA0DAAoBTj2GGfcLMzoBy+F0AOIAAAAA5EkgbG92ZSB5b3UsIGhvbmXheSEAwsFc%0D%0ABAABCgAQBQJWkFwRCRBOPYYZ9wszOgAA0tcQAItV6r329owQTvUngYS561fTay6R%0D%0AkUzbzvq2GKWcTFZ/sp0LoQH0i8HerfDYzWZ3rUCXdWvwggVqupGJEUPFTz2W4910%0D%0AIov6xTEs2/isvOHIenfmjx8HT25AdsXSnfpINFkRIJR7imHDfqNbcRQTOR8wRfLU%0D%0Ajoh3g+j1VGhSBpFQMVyWhyv2fA6cKm+rAZ0eJu+GcpOVF1R3wJWBh3eisQG9ZU2X%0D%0AU5xmjzRbPJxbzo66u7KNOJFRlpxGwIt2dqg4d1RIVq2CFIcGZoEJ0w8xtDml7WmX%0D%0ACVEz5Wy6ffP0n+Rh4MYJ/qN7mERq4mw4j2Qdhotf4/2zS7zT7bFw76IxKcI92Rzd%0D%0AA5RK/TrZNgC5ClEIPgHDw8rNlrh6cB+AYFyjCMtaqIKgy1aFoL3a5+F9wmmlD5tn%0D%0AJHPBiIJ7kPpx/GJ2W5gR8WV+iLfkWWUxQ9MGh/QUQueSd6QoXEbvRDtNEN4nY+Ln%0D%0AdnTC9Zq4AUL4f27Q0TNvM8nTUxPsDSYKG5CV5bAeTVjc7kBKvy8eADn6BsNUjAEk%0D%0A/SZGGC4CxMRGcu+xaVA1D3BwqvvUxI/zVLRuaDJ1akfsWK0lKSgUamvdwFsoOqYG%0D%0AV/DtqvxChx+4eQq4sow+jbqK93yRYuHM2i983WgdYzSxRdngwIIWKtp2otfeHSMT%0D%0Afc6ppa7rAmFsjnST%0D%0A=Y+Sh%0D%0A-----END%20PGP%20MESSAGE-----

signed by emilycoombsceo

The Signed Text

I love you, honey!

**Comment by malgorithms** [comment]
Thanks for writing all this up. We're pretty deep in the weeds of our desktop/mobile client right now, so the site isn't getting a lot of changes. 

For CORS we have the current policy:

  - it's only selectively available for a handful of GET requests on the site. We're unlikely to turn this on for any kind of POSTs or session-related activities for web security reasons. And we're unlikely to turn it on for static resources people can host themselves.   

Of course *any* non-web app can access all Keybase API calls (there's nothing about our own that is privileged), but the security concerns for web pages are very different; we're just talking about web/CORS here.

  - we don't support pre-flight OPTION requests, since last I understood, they aren't required for GET requests, and we haven't implemented them. Other people have shown me pages that do things like encrypt and verify using CORS calls, because you can look up all important information on a user (and search usrs) using CORS-supporting get requests on Keybase.

For example, this call supports CORS:

https://keybase.io/_/api/1.0/user/lookup.json?usernames=chris,max&fields=basics,public_keys
```
curl -I https://keybase.io/_/api/1.0/user/lookup.json?usernames=chris,max&fields=basics,public_keys

HTTP/1.1 200 OK
Date: Thu, 14 Jan 2016 15:58:02 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 143295
Connection: keep-alive
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET
Access-Control-Allow-Headers: Content-Type, Authorization, Content-Length, X-Requested-With
Access-Control-Allow-Credentials: false
...
```

Similarly, the raw key request works over CORS get requests, which I've seen people use:

https://keybase.io/chris/key.asc

In other words, I think you should be able to verify a message / reaffirm your wife's love using CORS as it is now, just don't do the preflight check, and stick to those calls.

**Comment by JasonCoombs** [comment]
Nice. Any reason we don't already support CORS for the keybase.io fetch API?

The following request is missing Access-Control-Allow-Origin: * which was half of the problem described above.

```
$ curl -I https://keybase.io/_/api/1.0/key/fetch.json?pgp_key_ids=4e3d8619f70b333a\&ops=4

HTTP/1.1 200 OK
Date: Thu, 14 Jan 2016 19:05:35 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 3686
Connection: keep-alive
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Set-Cookie: guest=lgHZIDk5ZTlmMGY0ZWI1NTRkYjcyOTAyMzNiYjRiM2ZjMjA4zlaX8X%2FOAAFRgMDEIJQCgTdxbBqCa5HaHGiY1AIyrwgbIgOFe3uVfGfeOKGG; Max-Age=604.8; Domain=.keybase.io; Path=/; Expires=Thu, 14 Jan 2016 19:15:40 GMT; HttpOnly; Secure
ETag: W/"e66-bDtyOn8NdQ1E1kYh1muwUA"
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
```

The other half of the problem described is an issue with the sitewide-js which, ideally, would become the basis of the web client widget API framework for keybase.io at least for encrypt, decrypt, sign and verify widgets. I agree that not all of the operations can be allowed without an active session, and there must be defenses against MITM attacks whenever private key passphrases are being entered by the user in cross-domain hosted widgets.

https://www.w3.org/TR/widgets/
https://www.w3.org/TR/widgets-apis/

RECOMMENDATIONS:

* Design and implement a full widget API for keybase.io
* Support CORS right away on the existing keybase.io fetch API just like the lookup API does already (this merely requires adding Access-Control-Allow-Origin: * to the fetch API server side code)
* Add code samples and instructions for scripting CORS requests from within different browsers and scripting frameworks -- the changes that I had to make to sitewide-js to stop preflight OPTIONS requests from happening are issues that arise from jquery or from RequireJS or whichever javascript framework it was that brought the jqXHR.setRequestHeader and options.crossDomain / X-Requested-With lines of code (removing them stopped preflight from happening on calls to the fetch API)
* Add a new keybase.io API for message processing, which of course should be compatible with CORS except where MITM and XSS are possible. For example, a msg/verify API could be added:

https://keybase.io/docs/api/1.0/call/msg/verify

The new msg/verify API would accept a BEGIN PGP MESSAGE block, either URL-encoded or POST'ed, and return either JSON or a widget-compatible browser-based user interface response depending on input parameters.

I like the fact that the existing sitewide-js includes a _run_if_prefilled() function and that it actually works (after the fetch API has been revised to support CORS and after the changes shown in my diff output have been made) but ideally msg/verify would be an API rather than just a widget functionality enabled through sitewide-js and its progeny.

Love keybase.io -- thanks for creating it!

**Comment by malgorithms** [comment]
> Nice. Any reason we don't already support CORS for the keybase.io fetch API?

It's paranoia / lack of confidence on my part. The fetch endpoint could be split in half, but right now it's the same endpoint from which users can request their encrypted private pgp key, if they use that feature. In theory browsers shouldn't let those CORS requests through because we set our CORS replies to have `Access-Control-Allow-Credentials: false`, but I fear that's not good enough. A browser bug, a misunderstanding on my part about CORS, or a bug we introduce in the headers could allow a site to let a user request their encrypted private key. So I just have CORS turned off for something that could provide sensitive info.

I could sometime split the fetch req up, but that brings us back to the fact that the public keys can be requested from another endpoint anyway.

> RECOMMENDATIONS:

all of those are good recommendations. We likely won't get to them soon due to our own client work.

**Comment by JasonCoombs** [comment]
Understood, thanks for walking me through that.

I will be able to further revise my copy of sitewide-js.js and/or its dependencies to switch from using the non-CORS-compatible fetch API as happens now for message verify:

https://keybase.io/_/api/1.0/key/fetch.json?pgp_key_ids=4e3d8619f70b333a&ops=4

To instead using the CORS-compatible lookup API with the full key fingerprint as the query parameter:

https://keybase.io/_/api/1.0/user/lookup.json?key_fingerprint=c7d189de24e556fb9553aac04e3d8619f70b333a

This will make my test verify page on jasoncoombs.com work without the CORS Chrome extension.

Will you consider creating a github project for the existing server-side code and website content for keybase.io so work could begin on adapting what is there now into a widget API and enhancements can be made to the website in general by outside contributors?

I would particularly like to be able to construct actual keybase.io/verify URLs (rather than using my test verify.php page at jasoncoombs.com) that will pre-fill messages so that _run_if_prefilled() will work as it was apparently intended at some point in the past during initial development. The best way to do this is not through the use of the client-side script line that I added to my verify page:

$("#input-msg").val(decodeURI(window.location.search).substr(1));

Rather, the best way to pre-fill the input field so that message verification will be done automatically is to revise the server-side script to parse and sanitize the URL search string, so that XSS attacks can be definitively prevented. My line of code, above, poses a clear XSS risk as the contents of window.location.search after the ? are simply copied into the $("#input-msg") element unverified.

Considering that you already have _run_if_prefilled() as part of sitewide-js perhaps there is already a server-side conditional check for such user input of the message to be pre-filled via URL parameter?

My immediate need for further prototype development involves only the end-user experience for message verify in a browser, but if you were to start a new github project for the keybase.io website and keybase API source code then I'd be able to start work on a msg/verify API that would output JSON and prototype ideas for a formal keybase.io widget. Thanks!

**Comment by patcon** [comment]
Seemingly related: https://github.com/keybase/keybase-issues/issues/2170
